pkg/endpoint/policy.go

```diff
@@ -311,41 +309,61 @@ func (e *Endpoint) regenerate(owner Owner) error {
 	return nil
 }
 
-// Force regeneration of endpoint programs & policy
-func (e *Endpoint) regenerateLocked(owner Owner) error {
-	err := e.regenerate(owner)
-	if err != nil {
-		e.LogStatus(BPF, Failure, err.Error())
-	} else {
-		e.LogStatusOK(BPF, "Successfully regenerated endpoint program")
-	}
-
-	return err
-}
-
-// Regenerate forces the regeneration of endpoint programs & policy.
-func (e *Endpoint) Regenerate(owner Owner) error {
-	return e.regenerateLocked(owner)
+// Regenerate forces the regeneration of endpoint programs & policy
+func (e *Endpoint) Regenerate(owner Owner) <-chan bool {
+	newReq := &Request{
+		ID:           e.ID,
+		MyTurn:       make(chan bool),
+		Done:         make(chan bool),
+		ExternalDone: make(chan bool),
+	}
+	owner.QueueEndpoint(newReq)
+	go func(e *Endpoint, myTurn <-chan bool, finish chan<- bool, externalFinish chan<- bool) {
```

Comment:
Nice catch, I changed the `StartEndpointBuilders` code a little bit:
```
...
go func() {
			for e := range d.buildEndpointChan {
				d.uniqueIDMU.Lock()
+				if _, ok := d.uniqueID[e.ID]; !ok {
+					// If the request is not present in the uniqueID,
+					// it means the request was deleted from the queue
+					// so we deny the request's turn.
+					e.MyTurn <- false
+					d.uniqueIDMU.Unlock()
+					continue
+				}
				// Set the endpoint to "building" state
				d.uniqueID[e.ID] = true
				e.MyTurn <- true
				d.uniqueIDMU.Unlock()
```

This way the goroutine sees that its turn was denied and ends cleanly.